src/main/scala/com/microsoft/hyperspace/index/covering/CoveringIndexTrait.scala

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (2021) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.covering
+
+import org.apache.spark.sql.{DataFrame, SaveMode}
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
+import org.apache.spark.sql.functions.col
+import org.apache.spark.sql.hyperspace.utils.StructTypeUtils
+import org.apache.spark.sql.types.StructType
+
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.util.ResolverUtils.ResolvedColumn
+
+/**
+ * CoveringIndexTrait is a common interface for covering index type which contains a vertical
+ * slice of source data including indexedColumns and includedColumns.
+ */
+trait CoveringIndexTrait extends Index {
+  val schema: StructType
+  val includedColumns: Seq[String]
```

Comment:
nit: I would rather not use `val` in traits unless there is a specific reason.

Refs:
1. https://alvinalexander.com/scala/fp-book/def-vs-val-abstract-members-traits-classes-override/
2. https://stackoverflow.com/questions/19642053/when-to-use-val-or-def-in-scala-traits

Reply:
Thanks! I'll address your comments in the next PR.